src/plugins/screenshot/Screenshot.cc

```diff
@@ -0,0 +1,130 @@
+
+#include <ignition/msgs/boolean.pb.h>
+#include <ignition/msgs/stringmsg.pb.h>
+
+#include <string>
+
+#include <ignition/common/Console.hh>
+#include <ignition/common/Filesystem.hh>
+#include <ignition/plugin/Register.hh>
+#include <ignition/transport/Node.hh>
+
+#include "ignition/gui/Application.hh"
+
+namespace ignition
+{
+namespace gui
+{
+namespace plugins
+{
+  class ScreenshotPrivate
+  {
+    /// \brief Node for communication
+    public: ignition::transport::Node node;
+
+    /// \brief Screenshot service name
+    public: std::string screenshotService;
+
+    /// \brief Directory to save screenshots
+    public: std::string directory;
+  };
+}
+}
+}
+
+using namespace ignition;
+using namespace gui;
+using namespace plugins;
+
+/////////////////////////////////////////////////
+Screenshot::Screenshot()
+  : ignition::gui::Plugin(),
+  dataPtr(std::make_unique<ScreenshotPrivate>())
+{
+  // for screenshot requests
+  this->dataPtr->screenshotService = "/gui/screenshot";
+
+  std::string home;
+  common::env(IGN_HOMEDIR, home);
+
+  // default directory
+  this->dataPtr->directory =
+      common::joinPaths(home, ".ignition", "gazebo", "pictures");
+
+  if (!common::exists(this->dataPtr->directory))
+  {
+    if (!common::createDirectory(this->dataPtr->directory))
+    {
+      std::string defaultDir = common::joinPaths(home, ".ignition", "gazebo");
+      ignerr << "Unable to create directory [" << this->dataPtr->directory
+             << "]. Changing default directory to: " << defaultDir
+             << std::endl;
+
+      this->dataPtr->directory = defaultDir;
+    }
+  }
+
+  // Update tooltip text
+  App()->Engine()->rootContext()->setContextProperty("ScreenshotDirectory",
+      QString::fromStdString(this->dataPtr->directory));
+}
+
+/////////////////////////////////////////////////
+Screenshot::~Screenshot() = default;
+
+/////////////////////////////////////////////////
+void Screenshot::LoadConfig(const tinyxml2::XMLElement *)
+{
+  if (this->title.empty())
+    this->title = "Screenshot";
+}
+
+/////////////////////////////////////////////////
+void Screenshot::OnScreenshot()
+{
+  std::function<void(const ignition::msgs::Boolean &, const bool)> cb =
+      [](const ignition::msgs::Boolean &/*_rep*/, const bool _result)
+  {
+    if (!_result)
+      ignerr << "Error sending move to request" << std::endl;
+  };
+
+  std::string time = common::systemTimeISO() + ".png";
+  std::string savedPath = common::joinPaths(this->dataPtr->directory, time);
+
+  ignition::msgs::StringMsg req;
+  req.set_data(savedPath);
+  this->dataPtr->node.Request(this->dataPtr->screenshotService, req, cb);
+}
+
+/////////////////////////////////////////////////
+void Screenshot::OnChangeDirectory(const QString &_dirUrl)
+{
+  QString newDir = QUrl(_dirUrl).toLocalFile();
+  this->dataPtr->directory = newDir.toStdString();
+
+  // Update tooltip text
+  App()->Engine()->rootContext()->setContextProperty("ScreenshotDirectory",
+      newDir);
```

Comment:
We should avoid using these "global" properties, because there could be collisions with other plugins. I suggest adding a property to this plugin.

See the `msgType` property of the `Publisher` plugin as an example:

https://github.com/ignitionrobotics/ign-gui/blob/9140459ed06a65e3045354064edfb7e4cc8b9e72/src/plugins/publisher/Publisher.hh#L51-L57

https://github.com/ignitionrobotics/ign-gui/blob/9140459ed06a65e3045354064edfb7e4cc8b9e72/src/plugins/publisher/Publisher.hh#L96-L107

Reply:
Another suggestion, you could even use the same approach to store the final timestamped path as a property, and display that on the dialog. That should also make the dialog only pop up after the image has actually been saved.

Reply:
Done 9b5e3db77216ffccca5a23b53fe101493e9f2f08